daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala

```diff
@@ -0,0 +1,430 @@
+  val basicTestsSignatures = language.PackageInterface(Map(basicTestsPkgId -> basicTestsPkg))
+
+  val withKeyTemplate = "BasicTests:WithKey"
+  val BasicTests_WithKey = Identifier(basicTestsPkgId, withKeyTemplate)
+  val withKeyContractInst: VersionedContractInstance =
+    assertAsVersionedContract(
+      ContractInstance(
+        TypeConName(basicTestsPkgId, withKeyTemplate),
+        ValueRecord(
+          Some(BasicTests_WithKey),
+          ImmArray(
+            (Some[Ref.Name]("p"), ValueParty(alice)),
+            (Some[Ref.Name]("k"), ValueInt64(42)),
+          ),
+        ),
+        "",
+      )
+    )
+
+  val defaultContracts: Map[ContractId, VersionedContractInstance] =
+    Map(
+      toContractId("BasicTests:Simple:1") ->
+        assertAsVersionedContract(
+          ContractInstance(
+            TypeConName(basicTestsPkgId, "BasicTests:Simple"),
+            ValueRecord(
+              Some(Identifier(basicTestsPkgId, "BasicTests:Simple")),
+              ImmArray((Some[Name]("p"), ValueParty(party))),
+            ),
+            "",
+          )
+        ),
+      toContractId("BasicTests:CallablePayout:1") ->
+        assertAsVersionedContract(
+          ContractInstance(
+            TypeConName(basicTestsPkgId, "BasicTests:CallablePayout"),
+            ValueRecord(
+              Some(Identifier(basicTestsPkgId, "BasicTests:CallablePayout")),
+              ImmArray(
+                (Some[Ref.Name]("giver"), ValueParty(alice)),
+                (Some[Ref.Name]("receiver"), ValueParty(bob)),
+              ),
+            ),
+            "",
+          )
+        ),
+      toContractId("BasicTests:WithKey:1") ->
+        withKeyContractInst,
+    )
+
+  val defaultKey = Map(
+    GlobalKey.assertBuild(
+      TypeConName(basicTestsPkgId, withKeyTemplate),
+      ValueRecord(None, ImmArray((None, ValueParty(alice)), (None, ValueInt64(42)))),
+    )
+      ->
+        toContractId("BasicTests:WithKey:1")
+  )
+
+  private[this] def lookupPackage(pkgId: PackageId): Option[Package] = {
+    allPackages.get(pkgId)
+  }
+
+  private[this] def lookupKey(key: GlobalKeyWithMaintainers): Option[ContractId] =
+    (key.globalKey.templateId, key.globalKey.key) match {
+      case (
+            BasicTests_WithKey,
+            ValueRecord(_, ImmArray((_, ValueParty(`alice`)), (_, ValueInt64(42)))),
+          ) =>
+        Some(toContractId("BasicTests:WithKey:1"))
+      case _ =>
+        None
+    }
+
+  private[this] val suffixLenientEngine = newEngine()
+  private[this] val preprocessor =
+    new preprocessing.Preprocessor(
+      ConcurrentCompiledPackages(suffixLenientEngine.config.getCompilerConfig)
+    )
+
+  "contract key" should {
+    val now = Time.Timestamp.now()
+    val submissionSeed = crypto.Hash.hashPrivateKey("contract key")
+    val txSeed = crypto.Hash.deriveTransactionSeed(submissionSeed, participant, now)
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated only when executing create
```

Comment:
```suggestion
    // TEST_EVIDENCE: Semantics: contract keys should be evaluated after ensure clause
```
maybe a bit clearer and matches what this test actually tests afaict.